tracee/pipeline.go

```diff
@@ -0,0 +1,231 @@
+
+	// Source pipeline stage.
+	rawEventChan, errc, err := t.decodeRawEvent(done)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	processedEventChan, errc, err := t.processRawEvent(done, rawEventChan)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	printEventChan, errc, err := t.prepareEventForPrint(done, processedEventChan)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	errc, err = t.printEvent(done, printEventChan)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	// Pipeline started. Waiting for pipeline to complete
+	return t.WaitForPipeline(errcList...)
+}
+
+type RawEvent struct {
+	Ctx      context
+	RawArgs  map[argTag]interface{}
+	ArgsTags []argTag
+}
+
+func (t *Tracee) decodeRawEvent(done <-chan struct{}) (<-chan RawEvent, <-chan error, error) {
+	out := make(chan RawEvent)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
+		for dataRaw := range t.eventsChannel {
+			dataBuff := bytes.NewBuffer(dataRaw)
+			var ctx context
+			err := binary.Read(dataBuff, binary.LittleEndian, &ctx)
+			if err != nil {
+				errc <- err
+				continue
+			}
+
+			rawArgs := make(map[argTag]interface{})
+			argsTags := make([]argTag, ctx.Argnum)
+			for i := 0; i < int(ctx.Argnum); i++ {
+				tag, val, err := readArgFromBuff(dataBuff)
+				if err != nil {
+					errc <- err
+					continue
+				}
+				argsTags[i] = tag
+				rawArgs[tag] = val
+			}
+			select {
+			case out <- RawEvent{ctx, rawArgs, argsTags}:
+			case <-done:
+				return
+			}
+		}
+	}()
+	return out, errc, nil
+}
+
+func (t *Tracee) processRawEvent(done <-chan struct{}, in <-chan RawEvent) (<-chan RawEvent, <-chan error, error) {
+	out := make(chan RawEvent)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
+		for rawEvent := range in {
+			err := t.processEvent(&rawEvent.Ctx, rawEvent.RawArgs)
+			if err != nil {
+				errc <- err
+				continue
+			}
+			select {
+			case out <- rawEvent:
+			case <-done:
+				return
+			}
+		}
+	}()
+	return out, errc, nil
+}
+
+func (t *Tracee) prepareEventForPrint(done <-chan struct{}, in <-chan RawEvent) (<-chan Event, <-chan error, error) {
+	out := make(chan Event, 1000)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
+		for rawEvent := range in {
+			if !t.shouldPrintEvent(rawEvent.Ctx.EventID) {
+				continue
+			}
+			err := t.prepareArgsForPrint(&rawEvent.Ctx, rawEvent.RawArgs)
+			if err != nil {
+				errc <- err
+				continue
+			}
+			args := make([]interface{}, rawEvent.Ctx.Argnum)
+			argsNames := make([]string, rawEvent.Ctx.Argnum)
+			for i, tag := range rawEvent.ArgsTags {
+				args[i] = rawEvent.RawArgs[tag]
+				argName, ok := argNames[tag]
+				if ok {
+					argsNames[i] = argName
+				} else {
+					errc <- fmt.Errorf("Invalid arg tag")
+					continue
+				}
+			}
+			t.stats.eventCounter.Increment()
```

Comment:
better to move that to printEvent (after the printing), to be sure that the counter is correct